File: vtrack-ios/VTrack/Views/Dashboard/DashboardScreen.swift
import SwiftUI
import SwiftData

struct DashboardScreen: View {
    @Environment(\.modelContext) private var modelContext
    @State private var viewModel: DashboardViewModel?

    var onAddVehicle: () -> Void
    var onNavigateToFuelEntry: (String) -> Void
    var onNavigateToMaintenanceTypes: () -> Void

    var body: some View {
        Group {
            if let vm = viewModel {
                if vm.vehicles.isEmpty {
                    EmptyStateView(
                        icon: "car.fill",
                        title: "Welcome to VTrack!",
                        message: "Add your first vehicle to get started",
                        actionTitle: "Add Vehicle",
                        action: onAddVehicle
                    )
                } else if vm.activeVehicle != nil {
                    ScrollView {
                        VStack(spacing: 16) {
                            if vm.vehicles.count > 1 {
                                vehicleSelector(vm: vm)
                            }
                            odometerCard(odometer: vm.currentOdometer)
                            lastFillUpCard(entry: vm.lastFuelEntry, mpg: vm.latestMpg)
                            maintenanceCard(statuses: vm.maintenanceStatuses)
                            totalSpendCard(total: vm.totalFuelCost)
                        }
                        .padding()
                    }
                } else {
                    ProgressView()
                }
            } else {
                ProgressView()
            }
        }
        .onAppear {
            if viewModel == nil {
                viewModel = DashboardViewModel(modelContext: modelContext)
            }
            viewModel?.load()
        }
    }

    @ViewBuilder
    private func vehicleSelector(vm: DashboardViewModel) -> some View {
        Menu {
            ForEach(vm.vehicles, id: \.persistentModelID) { vehicle in
                Button {
                    vm.selectVehicle(vehicle)
                } label: {
                    HStack {
                        Text(vehicle.displayName)
                        if vehicle.persistentModelID == vm.activeVehicle?.persistentModelID {
                            Image(systemName: "checkmark")
                        }
                    }
                }
            }
        } label: {
            HStack {
                Text(vm.activeVehicle?.displayName ?? "")
                    .font(.headline)
                Image(systemName: "chevron.down")
            }
            .frame(maxWidth: .infinity)
            .padding()
            .background(.thinMaterial)
            .clipShape(RoundedRectangle(cornerRadius: 12))
        }
    }

    @ViewBuilder
    private func odometerCard(odometer: Int) -> some View {
        VStack(spacing: 4) {
            Text(FormatUtil.formatMiles(odometer))
                .font(.title)
                .fontWeight(.bold)
            Text("Current Odometer")
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .frame(maxWidth: .infinity)
        .padding()
        .background(.regularMaterial)
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    @ViewBuilder
    private func lastFillUpCard(entry: FuelEntry?, mpg: Double?) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Last Fill-up")
                .font(.headline)
            if let entry {
                HStack {
                    Text(FormatUtil.formatDate(entry.date))
                    Spacer()
                    Text(FormatUtil.formatGallons(entry.gallons))
                    Text("•")
                    Text(FormatUtil.formatCurrency(entry.totalCost))
                }
                .font(.subheadline)
                if let mpg {
                    Text(FormatUtil.formatMpg(mpg))
                        .font(.subheadline)
                        .foregroundStyle(.secondary)
                }
            } else {
                Text("No fill-ups yet")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding()
        .background(.regularMaterial)
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    @ViewBuilder
    private func maintenanceCard(statuses: [MaintenanceStatus]) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Upcoming Maintenance")
                .font(.headline)
            let upcoming = statuses.filter { $0.urgency != .ok }
            if upcoming.isEmpty {
                HStack {
                    Image(systemName: "checkmark.circle.fill")
                        .foregroundStyle(.green)
                    Text("All clear!")
                }
                .font(.subheadline)
            } else {
                ForEach(upcoming, id: \.type.persistentModelID) { status in
                    HStack {
                        Image(systemName: status.urgency == .overdue ? "exclamationmark.triangle.fill" : "exclamationmark.circle.fill")
                            .foregroundStyle(status.urgency == .overdue ? .red : .orange)
                        VStack(alignment: .leading) {
                            Text(status.type.name)
                                .font(.subheadline)
                                .fontWeight(.medium)
                            Text(status.urgency == .overdue
                                 ? "Overdue by \(FormatUtil.formatMiles(abs(status.milesUntilDue)))"
                                 : "\(FormatUtil.formatMiles(status.milesUntilDue)) remaining")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                        }
                        Spacer()
                    }
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding()
        .background(.regularMaterial)
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    @ViewBuilder
    private func totalSpendCard(total: Double) -> some View {
        VStack(spacing: 4) {
            Text(FormatUtil.formatCurrency(total))
                .font(.title2)
                .fontWeight(.bold)
            Text("Total Fuel Spend")
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .frame(maxWidth: .infinity)
        .padding()
        .background(.regularMaterial)
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

File: app/src/main/java/com/vtrack/feature/dashboard/DashboardScreen.kt
package com.vtrack.feature.dashboard

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ElevatedCard
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.MenuAnchorType
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.navigation.NavController
import com.vtrack.navigation.Route
import com.vtrack.util.FormatUtil
import com.vtrack.util.MaintenanceStatus
import com.vtrack.util.MaintenanceUrgency

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun DashboardScreen(
    navController: NavController,
    viewModel: DashboardViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsState()

    when {
        uiState.isLoading -> {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator()
            }
        }

        uiState.hasNoVehicle -> {
            EmptyVehicleState(
                onAddVehicle = {
                    navController.navigate(Route.VehicleForm.create())
                }
            )
        }

        else -> {
            DashboardContent(
                uiState = uiState,
                onSwitchVehicle = { viewModel.switchVehicle(it) }
            )
        }
    }
}

@Composable
private fun EmptyVehicleState(onAddVehicle: () -> Unit) {
    Box(
        modifier = Modifier.fillMaxSize(),
        contentAlignment = Alignment.Center
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
            modifier = Modifier.padding(32.dp)
        ) {
            Icon(
                imageVector = Icons.Default.Add,
                contentDescription = null,
                modifier = Modifier.size(72.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(16.dp))
            Text(
                text = "Welcome to Autometer!",
                style = MaterialTheme.typography.headlineSmall,
                color = MaterialTheme.colorScheme.onSurface,
                textAlign = TextAlign.Center
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = "Add your first vehicle to get started",
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
            Spacer(modifier = Modifier.height(24.dp))
            Button(onClick = onAddVehicle) {
                Icon(
                    imageVector = Icons.Default.Add,
                    contentDescription = null,
                    modifier = Modifier.size(18.dp)
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text("Add Vehicle")
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun DashboardContent(
    uiState: DashboardUiState,
    onSwitchVehicle: (Long) -> Unit
) {
    LazyColumn(
        modifier = Modifier.fillMaxSize(),
        contentPadding = PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        if (uiState.vehicles.size > 1) {
            item(key = "vehicle_selector") {
                VehicleSelector(
                    vehicles = uiState.vehicles,
                    selectedVehicle = uiState.vehicle,
                    onVehicleSelected = onSwitchVehicle
                )
            }
        }

        item(key = "odometer_card") {
            OdometerCard(
                currentOdometer = uiState.currentOdometer,
                totalMiles = uiState.totalMiles
            )
        }

        item(key = "last_fillup_card") {
            LastFillUpCard(uiState = uiState)
        }

        item(key = "fuel_economy_card") {
            FuelEconomyCard(averageMpg = uiState.averageMpg)
        }

        item(key = "maintenance_card") {
            MaintenanceCard(statuses = uiState.upcomingMaintenance)
        }

        item(key = "total_spend_card") {
            TotalSpendCard(totalSpend = uiState.totalFuelSpend)
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun VehicleSelector(
    vehicles: List<com.vtrack.data.model.Vehicle>,
    selectedVehicle: com.vtrack.data.model.Vehicle?,
    onVehicleSelected: (Long) -> Unit
) {
    var expanded by remember { mutableStateOf(false) }

    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = it }
    ) {
        OutlinedTextField(
            value = selectedVehicle?.let { "${it.year} ${it.make} ${it.model}" } ?: "",
            onValueChange = {},
            readOnly = true,
            label = { Text("Active Vehicle") },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            modifier = Modifier
                .menuAnchor(MenuAnchorType.PrimaryNotEditable)
                .fillMaxWidth()
        )
        ExposedDropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false }
        ) {
            vehicles.forEach { vehicle ->
                DropdownMenuItem(
                    text = { Text("${vehicle.year} ${vehicle.make} ${vehicle.model}") },
                    onClick = {
                        onVehicleSelected(vehicle.id)
                        expanded = false
                    },
                    leadingIcon = if (vehicle.id == selectedVehicle?.id) {
                        {
                            Icon(
                                imageVector = Icons.Default.Check,
                                contentDescription = null
                            )
                        }
                    } else null
                )
            }
        }
    }
}

@Composable
private fun OdometerCard(currentOdometer: Int, totalMiles: Int) {
    ElevatedCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "Odometer",
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = FormatUtil.formatMiles(currentOdometer),
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.onSurface
            )
            Spacer(modifier = Modifier.height(4.dp))
            Text(
                text = "${FormatUtil.formatMiles(totalMiles)} tracked",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

@Composable
private fun LastFillUpCard(uiState: DashboardUiState) {
    ElevatedCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "Last Fill-up",
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(8.dp))
            if (uiState.lastFillUp != null) {
                val entry = uiState.lastFillUp
                Text(
                    text = FormatUtil.formatDate(entry.date),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onSurface
                )
                Spacer(modifier = Modifier.height(4.dp))
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Text(
                        text = FormatUtil.formatGallons(entry.gallons),
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    Text(
                        text = FormatUtil.formatCurrency(entry.totalCost),
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.SemiBold,
                        color = MaterialTheme.colorScheme.onSurface
                    )
                }
                if (uiState.lastMpg != null) {
                    Spacer(modifier = Modifier.height(4.dp))
                    Text(
                        text = FormatUtil.formatMpg(uiState.lastMpg),
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.primary
                    )
                }
            } else {
                Text(
                    text = "No fill-ups yet",
                    style = MaterialTheme.typography.bodyLarge,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

@Composable
private fun FuelEconomyCard(averageMpg: Double?) {
    ElevatedCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "Fuel Economy",
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = if (averageMpg != null) FormatUtil.formatMpg(averageMpg) else "\u2014",
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.Bold,
                color = if (averageMpg != null) {
                    MaterialTheme.colorScheme.primary
                } else {
                    MaterialTheme.colorScheme.onSurfaceVariant
                }
            )
            if (averageMpg != null) {
                Text(
                    text = "Average",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

@Composable
private fun MaintenanceCard(statuses: List<MaintenanceStatus>) {
    ElevatedCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "Upcoming Maintenance",
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(8.dp))
            if (statuses.isEmpty()) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        imageVector = Icons.Default.Check,
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.primary,
                        modifier = Modifier.size(20.dp)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        text = "All clear!",
                        style = MaterialTheme.typography.bodyLarge,
                        color = MaterialTheme.colorScheme.primary
                    )
                }
            } else {
                statuses.forEach { status ->
                    MaintenanceStatusRow(status = status)
                    Spacer(modifier = Modifier.height(8.dp))
                }
            }
        }
    }
}

@Composable
private fun MaintenanceStatusRow(status: MaintenanceStatus) {
    val urgencyColor = when (status.urgency) {
        MaintenanceUrgency.OVERDUE -> MaterialTheme.colorScheme.error
        MaintenanceUrgency.DUE_SOON -> MaterialTheme.colorScheme.tertiary
        MaintenanceUrgency.OK -> MaterialTheme.colorScheme.onSurfaceVariant
    }

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.weight(1f)
        ) {
            Icon(
                imageVector = Icons.Default.Warning,
                contentDescription = null,
                tint = urgencyColor,
                modifier = Modifier.size(18.dp)
            )
            Spacer(modifier = Modifier.width(8.dp))
            Text(
                text = status.type.name,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface
            )
        }
        Text(
            text = if (status.milesUntilDue > 0) {
                "${FormatUtil.formatMiles(status.milesUntilDue)} left"
            } else {
                "Overdue by ${FormatUtil.formatMiles(-status.milesUntilDue)}"
            },
            style = MaterialTheme.typography.bodySmall,
            fontWeight = FontWeight.SemiBold,
            color = urgencyColor
        )
    }
}

@Composable
private fun TotalSpendCard(totalSpend: Double) {
    ElevatedCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "Total Fuel Spend",
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = FormatUtil.formatCurrency(totalSpend),
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.onSurface
            )
        }
    }
}
